src/Microsoft.AspNetCore.Server.Kestrel/Adapter/Internal/RawStream.cs

```diff
@@ -126,9 +111,31 @@ public override Task FlushAsync(CancellationToken cancellationToken)
             return TaskCache.CompletedTask;
         }
 
-        private ValueTask<int> ReadAsync(ArraySegment<byte> buffer)
+        private async Task<int> ReadAsync(ArraySegment<byte> buffer)
         {
-            return _input.ReadAsync(buffer.Array, buffer.Offset, buffer.Count);
+            while (true)
+            {
+                var result = await _input.ReadAsyncDispatched();
+                var readableBuffer = result.Buffer;
+                try
+                {
+                    if (!result.Buffer.IsEmpty)
+                    {
+                        var count = Math.Min(result.Buffer.Length, buffer.Count);
+                        readableBuffer = readableBuffer.Slice(0, count);
+                        readableBuffer.CopyTo(buffer);
+                        return count;
+                    }
+                    else if (result.IsCompleted || result.IsCancelled)
+                    {
+                        return 0;
+                    }
+                }
+                finally
+                {
+                    _input.AdvanceReader(readableBuffer.End, readableBuffer.End);
```

Comment:
If an exception is thrown but `readableBuffer` has already been reassigned, wouldn't you want this to advance the reader till the end of `result.Buffer` (i.e. not just the end of the slice)?